service/resource/certificate/update_test.go

```diff
@@ -1,13 +1,304 @@
 package certificate
 
 import (
+	"context"
+	"reflect"
 	"testing"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/spf13/afero"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/pkg/api/v1"
+
+	"github.com/giantswarm/micrologger/microloggertest"
 )
 
 // Test_Resource_Certificate_GetUpdateState tests the GetUpdateState method.
 func Test_Resource_Certificate_GetUpdateState(t *testing.T) {
+	tests := []struct {
+		currentState []certificateFile
+		desiredState []certificateFile
+
+		expectedUpdateStateCertificateFiles []certificateFile
+		expectedErrorHandler                func(error) bool
+	}{
+		// Test that when the current state and desired state are both nil,
+		// a nil update state is returned.
+		{
+			currentState: nil,
+			desiredState: nil,
+
+			expectedUpdateStateCertificateFiles: nil,
+			expectedErrorHandler:                nil,
+		},
+
+		// Test that when the current state is nil, and desired state is empty,
+		// a nil update state is returned.
+		{
+			currentState: nil,
+			desiredState: []certificateFile{},
+
+			expectedUpdateStateCertificateFiles: nil,
+			expectedErrorHandler:                nil,
+		},
+
+		// Test that when the current state is empty, and desired state is nil,
+		// a nil update state is returned.
+		{
+			currentState: []certificateFile{},
+			desiredState: nil,
+
+			expectedUpdateStateCertificateFiles: nil,
+			expectedErrorHandler:                nil,
+		},
+
+		// Test that when the current and desired state are both empty,
+		// a nil update state is returned.
+		{
+			currentState: []certificateFile{},
+			desiredState: []certificateFile{},
+
+			expectedUpdateStateCertificateFiles: nil,
+			expectedErrorHandler:                nil,
+		},
+
+		// Test that when the current state is empty,
+		// and the desired state contains a certificate file,
+		// the update state contains the certificate file.
+		{
+			currentState: []certificateFile{},
+			desiredState: []certificateFile{
+				{
+					path: "/certs/foo",
+					data: "bar",
+				},
+			},
+
+			expectedUpdateStateCertificateFiles: []certificateFile{
+				{
+					path: "/certs/foo",
+					data: "bar",
+				},
+			},
+			expectedErrorHandler: nil,
+		},
+
+		// Test that when the current state contains a certificate,
+		// and the desired state contains the same certificate,
+		// the update state is nil.
+		{
+			currentState: []certificateFile{
+				{
+					path: "/certs/foo",
+					data: "bar",
+				},
+			},
+			desiredState: []certificateFile{
+				{
+					path: "/certs/foo",
+					data: "bar",
+				},
+			},
+
+			expectedUpdateStateCertificateFiles: nil,
+			expectedErrorHandler:                nil,
+		},
+
+		// Test that when the current state contains a certificate,
```

Comment:
Let's please have numbers here since there are quite a few tests. It helps me navigate when tests break.